source/common/rate_limiter_impl.cc

```diff
@@ -78,6 +78,43 @@ void LinearRateLimiter::releaseOne() {
   acquired_count_--;
 }
 
+LinearRampingRateLimiterImpl::LinearRampingRateLimiterImpl(Envoy::TimeSource& time_source,
+                                                           const std::chrono::nanoseconds ramp_time,
+                                                           const Frequency frequency)
+    : RateLimiterBaseImpl(time_source), ramp_time_(ramp_time), frequency_(frequency) {
+  if (frequency_.value() <= 0) {
+    throw NighthawkException("frequency must be > 0");
+  }
+  if (ramp_time <= 0ns) {
+    throw NighthawkException("duration must be positive");
+  }
+}
+
+bool LinearRampingRateLimiterImpl::tryAcquireOne() {
+  if (acquireable_count_) {
+    acquired_count_++;
+    return acquireable_count_--;
+  }
+  const auto elapsed_time = elapsed() + 1ns;
+  double elapsed_fraction = 1.0;
+  if (elapsed_time < ramp_time_) {
+    elapsed_fraction -= static_cast<double>(ramp_time_.count() - elapsed_time.count()) /
+                        static_cast<double>(ramp_time_.count());
+  }
+
+  const double current_frequency = elapsed_fraction * (frequency_.value() * 1.0);
+  const double chrono_seconds =
+      std::chrono::duration_cast<std::chrono::duration<double>>(elapsed_time).count();
+  const double total = chrono_seconds * current_frequency / 2.0;
```

Comment:
Sorry I may be tired, can you help me understand the division by 2.0?

Reply:
Added comment in 6329377a31b0b35828d81c2e2a22cfd8f3943e8a

Reply:
Thank you, that makes perfect sense.